File: 2289.steps-to-make-array-non-decreasing.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/*
 * @lc app=leetcode id=2289 lang=java
 *
 * [2289] Steps to Make Array Non-decreasing
 */

// @lc code=start
class Solution {
    public int totalSteps(int[] nums) {
        int n = nums.length;
        int[] arr = new int[n];
        Arrays.fill(arr, 0);
        int[] stack = new int[n];
        int top = -1, result = 0;
        for (int i = n - 1; i >= 0; i--) {
            while (top >= 0 && nums[i] > nums[stack[top]]) {
                arr[i] = Math.max(++arr[i], arr[stack[top--]]);
                result = Math.max(result, arr[i]);
            }
            stack[++top] = i;
        }

        return result;
    }
}
// @lc code=end

File: python/2289.steps-to-make-array-non-decreasing.py
#
# @lc app=leetcode id=2289 lang=python3
#
# [2289] Steps to Make Array Non-decreasing
#

# @lc code=start
from typing import List


class Solution:
    def totalSteps(self, nums: List[int]) -> int:
        n = len(nums)
        arr = [0] * n
        stack = []
        for i in range(n-1, -1, -1):
            while stack and nums[i] > nums[stack[-1]]:
                arr[i] = max(arr[i] + 1, arr[stack.pop()])
            stack.append(i)
        return max(arr)


obj = Solution()
print(obj.totalSteps([5, 3, 4, 4, 7, 3, 6, 11, 8, 5, 11]))

# @lc code=end
